internal/controller/resource_manager.go

```diff
@@ -63,9 +65,9 @@ func NewResourceManager(config ResourceManagerConfig) *ResourceManager {
 
 // WatchK8sResources initiates a watch stream for the specified Kubernetes resource starting from the given resourceVersion.
 // This function blocks until the watch ends or the context is canceled.
-func (r *ResourceManager) WatchK8sResources(ctx context.Context, cancel context.CancelFunc, apiGroup string, resourceVersion string) {
+func (r *ResourceManager) WatchK8sResources(ctx context.Context, cancel context.CancelFunc, apiGroup string, resourceVersion string, mutationChan chan *pb.KubernetesResourceMutation, watcherWaitGroup *sync.WaitGroup) {
 	defer cancel()
-
+	defer watcherWaitGroup.Done()
```

Comment:
Revert adding the `WaitGroup` as a parameter. Do that logic in the calling goroutine function.
```suggestion
func (r *ResourceManager) WatchK8sResources(ctx context.Context, cancel context.CancelFunc, apiGroup string, resourceVersion string, mutationChan chan *pb.KubernetesResourceMutation) {
	defer cancel()
```